### PR TITLE
Update shamie-Resnet.mdx

### PR DESCRIPTION
**Pull request by ShamieCC**
Resnet summary, architecture, and code.

### CODE REVIEW

**Review by sezan92 — COMMENTED** [review]
i will review tommorrow or the day after

File: shamie-Resnet.mdx [hunk 1]
```diff
@@ -1,24 +1,126 @@
...
+### Observed Problem: Degradation with depth.
 
+Deeper network layers were assumed to be more effective because the performance of neural networks improved by deepening the layers.
 As the layers deepened, the extracted features could be further enriched, such as seen with VGG16 and VGG19.
+A question arose: "Is learning networks as easy as stacking more layers"?
+An obstacle to answering this question, the gradient vanishing problem, was addressed by normalized initializations and intermediate normalization layers.
+
+However, a new issue emerged: the degradation problem. As the neural networks deepened, accuracy saturated and degraded rapidly. An experiment comparing shallow and deep plain networks revealed that deeper models exhibited higher training and test errors, suggesting a fundamental challenge in training deeper architectures effectively. This degradation was not because of overfitting but because the training error increased when the network deepened. The added layers did not approximate the identity function.
```
Comment (sezan92) [inline shamie-Resnet.mdx]: can you give reference to the experiment?

File: shamie-Resnet.mdx [hunk 2]
```diff
@@ -1,24 +1,126 @@
...
+
+Resnet’s residual connections unlocked the potential of extreme depth, propelling accuracy upwards while plain networks grappled with degradation.
```
Comment (sezan92) [inline shamie-Resnet.mdx]:
```suggestion
Resnet’s residual connections unlocked the potential of the extreme depth, propelling the accuracy upwards compared to the previous architectures.

```

File: shamie-Resnet.mdx [hunk 3]
```diff
@@ -1,24 +1,126 @@
...
+
+## Resnet Architecture
+
+- A Residual Block. Source: ResNet Paper
```
Comment (sezan92) [inline shamie-Resnet.mdx]: reference of the paper?

File: shamie-Resnet.mdx [hunk 4]
```diff
@@ -1,24 +1,126 @@
...
+
+### H(x) = x
+
+Shortcut connections perform identity mapping and their output is added to the output of the stacked layers. Identity shortcut connections add neither extra parameters nor
+computational complexity, these connections bypass layers, creating direct paths for information flow, and they enable neural networks to learn the residual function (F).
+
+Resnet Network = Plain Network + Shortcuts
+
+For operations F(x) + x, F(x) and x should have identical dimensions.
+ResNet employs two techniques to achieve this:
```
Comment (sezan92) [inline shamie-Resnet.mdx]: @sitamgithub-MSIT do you have a special diagram for this?

File: shamie-Resnet.mdx [hunk 5]
```diff
@@ -1,24 +1,126 @@
...
+
+- Zero-padding shortcuts that add channels with zero values, maintaining dimensions without introducing extra parameters to be learned.
+- Projection shortcuts that use 1x1 convolutions to adjust dimensions when necessary, involving some additional learnable parameters.
+
+In deeper ResNet architectures like ResNet 50, 101, and 152, a specialized "bottleneck building block" is employed to manage parameter complexity and maintain efficiency while enabling even deeper learning.
```
Comment (sezan92) [inline shamie-Resnet.mdx]: can you add one or two lines about deeper Resnet architectures ? i mean the problem with the deeper networks

File: shamie-Resnet.mdx [hunk 6]
```diff
@@ -1,24 +1,126 @@
...
+
+## Resnet Code
+
+### Author: Pytorch Team
```
Comment (sezan92) [inline shamie-Resnet.mdx]: this part can be made as docstring

File: shamie-Resnet.mdx [hunk 7]
```diff
@@ -1,24 +1,126 @@
...
+
+### Deep residual networks pre-trained on ImageNet
+
+import torch
```
Comment (sezan92) [inline shamie-Resnet.mdx]:
```suggestion
```
import torch

```

Comment (sezan92) [inline shamie-Resnet.mdx]: add "```" before the `import torch`

File: shamie-Resnet.mdx [hunk 8]
```diff
@@ -1,24 +1,126 @@
...
+# Read the categories
+
+with open("imagenet_classes.txt", "r") as f:
+categories = [s.strip() for s in f.readlines()]
+
+# Show top categories per image
+
+top5_prob, top5_catid = torch.topk(probabilities, 5)
+for i in range(top5_prob.size(0)):
+print(categories[top5_catid[i]], top5_prob[i].item())
```
Comment (sezan92) [inline shamie-Resnet.mdx]: add "```" after the print statement

**Review by sezan92 — CHANGES_REQUESTED** [review]
commented. i feel there are some grammatical errors. can you recheck with Grammarly?

File: shamie-Resnet.mdx [hunk 9]
```diff
@@ -1,24 +1,126 @@
...
+import torch
+model = torch.hub.load('pytorch/vision:v0.10.0', 'resnet18', pretrained=True)
+# or any of these variants
+# model = torch.hub.load('pytorch/vision:v0.10.0', 'resnet34', pretrained=True)
+# model = torch.hub.load('pytorch/vision:v0.10.0', 'resnet50', pretrained=True)
+# model = torch.hub.load('pytorch/vision:v0.10.0', 'resnet101', pretrained=True)
+# model = torch.hub.load('pytorch/vision:v0.10.0', 'resnet152', pretrained=True)
+model.eval()
+
+All pre-trained models expect input images normalized in the same way, i.e. mini-batches of 3-channel RGB images of shape (3 x H x W), where H and W are expected to be at least 224. The images have to be loaded in to a range of [0, 1] and then normalized using mean = [0.485, 0.456, 0.406] and std = [0.229, 0.224, 0.225].
```
Comment (sitammeur) [inline shamie-Resnet.mdx]:
```suggestion
All pre-trained models expect input images normalized similarly, i.e. mini-batches of 3-channel RGB images of shape (3 x H x W), where H and W are expected to be at least 224. The images have to be loaded into a range of [0, 1] and then normalized using mean = [0.485, 0.456, 0.406] and std = [0.229, 0.224, 0.225].

```

File: shamie-Resnet.mdx [hunk 10]
```diff
@@ -1,24 +1,126 @@
...
+
+Here's a sample execution.
+
+# Download an example image from the pytorch website
```
Comment (sitammeur) [inline shamie-Resnet.mdx]:
```suggestion
# Download an example image from the Pytorch website

```

File: shamie-Resnet.mdx (hunk 4 above)
Comment (sitammeur) [reply shamie-Resnet.mdx]: Yeah, I suppose I had one regarding it.

**Review by sitammeur — CHANGES_REQUESTED** [review]
Everything is almost okay. Now, I think we should finalise it with PR. Also, if possible, can you try the RESNET code block, not importing from Torchvision but rather the implementation of the backbone?

File: shamie-Resnet.mdx [hunk 11]
```diff
@@ -1,24 +1,128 @@
...
+
+In deeper ResNet architectures like ResNet 50, 101, and 152, a specialized "bottleneck building block" is employed to manage parameter complexity and maintain efficiency while enabling even deeper learning.
+
+## Resnet Code
+
+### Author: Pytorch Team
+
+### Deep residual networks pre-trained on ImageNet
+
+import torch
```
Comment (sezan92) [inline shamie-Resnet.mdx]:
```suggestion

```
import torch

```

File: shamie-Resnet.mdx [hunk 12]
```diff
@@ -1,24 +1,129 @@
...
+# Read the categories
+
+with open("imagenet_classes.txt", "r") as f:
+categories = [s.strip() for s in f.readlines()]
+
+# Show top categories per image
+
+top5_prob, top5_catid = torch.topk(probabilities, 5)
+for i in range(top5_prob.size(0)):
+print(categories[top5_catid[i]], top5_prob[i].item())
```
Comment (sezan92) [inline shamie-Resnet.mdx]:
```suggestion
print(categories[top5_catid[i]], top5_prob[i].item())
```
```
